feat: add docker support

JefvdA opened the pull request:
I added a `Dockerfile` and `docker-compose.yaml`.

I almost never install software locally, and thus I added docker support for your project. If this is not something you want, no worries. Also if something is not clear, or you want me to change something before implementing the changes, let me know.

craftycodie commented:
Just a nit-pick, other than that it looks. great! Thank you :)

JefvdA commented:
> Just a nit-pick, other than that it looks. great! Thank you :)

Nono, no nit-pick. I forgot to ask this as well, as I was not sure what version to use. I'll change it 👍

craftycodie commented:
I had no idea mongo-express was a thing, looks useful.

SarahGreyWolf commented on docker-compose.yaml:
```diff
@@ -0,0 +1,58 @@
+      - main
+
+  mongo:
+    image: mongo
```
It's not ideal but making mongo use version 4.4.6 will allow it to run on ARM64/v8 systems like the raspberry pi 4.
Any newer version than that doesn't work thanks to mongo targeting a slightly newer ARM64/v8 architecture than these devices

JefvdA replied:
I updated the docker-compose to now use mongo version 4.4.6. Do you think I should change versions of other docker images aswell, I am not quite sure if they support ARM64/v8 architecture aswell

craftycodie reviewed (changes requested):
Please update the readme to explain docker setup, doesn't have to be too in-depth :)

SarahGreyWolf commented:
The Dockerfile doesn't really seem necessary at all since you never use nest-cli.
Also running stuff in containers using commands is not really the right way to use docker, it should be mostly contained within the Dockerfile for the image, with stuff in docker-compose used for configuring the app.

JefvdA commented on Dockerfile:
```diff
@@ -0,0 +1,13 @@
+FROM node:19.7-alpine AS builder
```
I've updated the Dockerfile to now use the LTS version 18.15

JefvdA commented:
> This is currently missing the required `MONGO_URI` variable

You're right, the node service shouldn't rely on the `MONGO_URI` from within the .env file. I added this variable to the node service, and set the `MONGO_URI` to the uri of the mongo container.

craftycodie commented on Dockerfile:
```diff
@@ -0,0 +1,13 @@
+ARG NPM_CONFIG_PREFIX=/.npm-global
+RUN npm install -g @nestjs/cli
+
+FROM node:19.7-alpine AS dev
```
Is this used?

JefvdA replied:
Sarah said this aswell, but the nest-cli is used for starting the server, that's why I included it. I don't know why I installed it globally tho, as it's actually in the `devDependencies` aswell. I'll see if I can just ditch the Dockerfile all together.

JefvdA replied:
This is the reason I added the Dockerfile with the global install for `nest-cli` was because `npm run start` uses nest to start the application. The problem is that if you use `nest start` as a command, you will need the nest-cli installed locally. I updated the docker-compose to now use the local installation of the nest-cli instead, this way I can just remove the Dockerfile. 

Previous start command: `npm run start`
New start command: `node_modules/.bin/nest start`

JefvdA commented:
> The Dockerfile doesn't really seem necessary at all since you never use nest-cli. Also running stuff in containers using commands is not really the right way to use docker, it should be mostly contained within the Dockerfile for the image, with stuff in docker-compose used for configuring the app.

You were right about the Dockerfile, I removed the commands depending on a global install of `nest-cli` and thus was able to remove the Dockerfile. [the conversation](https://github.com/craftycodie/Xenia-WebServices/pull/3#discussion_r1133078881)

As for the commands being ran in the containers, is now needed because I removed the Dockerfile. Do you think I should re-add the Dockerfile, so that I can remove the commands from the docker-compose?

Margen67 commented on docker-compose.yaml:
```diff
@@ -0,0 +1,62 @@
+      ME_CONFIG_MONGODB_ADMINPASSWORD: root
+      ME_CONFIG_MONGODB_URL: mongodb://root:root@mongo:27017/
+    networks:
+      - main
```
End of file newline
